aspects/aspects.go

```diff
@@ -0,0 +1,247 @@
+			}
+
+			access := aspectView["access"]
+			if access != "" && strutil.ListContains(validAccessValues, access) {
+				return nil, fmt.Errorf("cannot create aspect view: expected \"access\" to be one of %s instead of %q",
+					strutil.Quoted(validAccessValues), access)
+			}
+
+			aspect.accessPatterns = append(aspect.accessPatterns, &accessPattern{
+				name:   name,
+				path:   path,
+				access: access,
+			})
+		}
+
+		aspectDir.aspects[name] = aspect
+	}
+
+	return &aspectDir, nil
+}
+
+// Aspect return an aspect from the aspect directory.
+func (d *Directory) Aspect(aspect string) *Aspect {
+	return d.aspects[aspect]
+}
+
+// Aspect is a group of access patterns under a directory.
+type Aspect struct {
+	Name           string
+	accessPatterns []*accessPattern
+	directory      Directory
+}
+
+func (a *Aspect) Set(name string, value interface{}) error {
+	// TODO: add access control; name validation
+	for _, p := range a.accessPatterns {
+		if p.name == name {
+			if err := a.directory.dataBag.Set(p.path, value); err != nil {
+				return err
+			}
+
+			data, err := a.directory.dataBag.Data()
+			if err != nil {
+				return err
+			}
+
+			return a.directory.schema.Validate(data)
+		}
+	}
+
+	return &ErrNotFound{fmt.Sprintf("cannot set name %q in aspect %q: access pattern not found", name, a.Name)}
+}
+
+func (a *Aspect) Get(name string, value interface{}) error {
+	// TODO: add access control; name validation
+	for _, p := range a.accessPatterns {
+		if p.name == name {
+			return a.directory.dataBag.Get(p.path, value)
+		}
+	}
+
+	return &ErrNotFound{fmt.Sprintf("cannot get name %q in aspect %q: access pattern not found", name, a.Name)}
+}
+
+// accessPattern is an aspect accessPattern that holds information about a accessPattern.
+type accessPattern struct {
+	name   string
+	path   string
+	access string
+}
+
+// JSONDataBag is a simple DataBag implementation that keeps JSON in-memory.
+type JSONDataBag map[string]json.RawMessage
+
+func NewJSONDataBag() JSONDataBag {
+	storage := make(map[string]json.RawMessage)
+	return storage
+}
+
+func (s JSONDataBag) Get(path string, value interface{}) error {
+	subKeys := strings.Split(path, ".")
+	return get(subKeys, s, value)
+}
+
+func (s JSONDataBag) Set(path string, value interface{}) error {
+	subKeys := strings.Split(path, ".")
+	_, err := set(subKeys, s, value)
+	return err
+}
+
+func (s JSONDataBag) Data() ([]byte, error) {
+	return json.Marshal(s)
+}
+
+type JSONSchema struct{}
+
+func NewJSONSchema() *JSONSchema {
+	return &JSONSchema{}
+}
+
+func (s *JSONSchema) Validate(jsonData []byte) error {
+	// the top-level is always an object
+	var data map[string]json.RawMessage
+	return json.Unmarshal(jsonData, &data)
+}
+
+func get(subKeys []string, root map[string]json.RawMessage, result interface{}) error {
+	key := subKeys[0]
+	value, ok := root[key]
+	if !ok {
+		return &ErrNotFound{"key not found"}
+	}
+
+	if len(subKeys) == 1 {
+		return json.Unmarshal(value, result)
+	}
+
```

Comment:
we might need a way to detect if the next level is a map or something else to produce a cleaner error?